sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch/reminders/ReminderManager.kt

```diff
@@ -0,0 +1,251 @@
+        val reminderJson = gson.toJson(reminder)
+        pendingIntentPrefs(context).edit().putString(reminder.guid, reminderJson).apply()
+
+        reminder.repeatAlarmInterval?.let {
+            logger.info("Setting repeat reminder with info $reminder")
+            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, reminder.alarmEpochMillis, it, pendingIntent)
+        } ?: run {
+            logger.info("Setting reminder with info $reminder")
+            alarmManager.set(AlarmManager.RTC_WAKEUP, reminder.alarmEpochMillis, pendingIntent)
+        }
+    }
+
+    /**
+     * Cancels a single reminder
+     * @param context can be app or activity
+     * @param reminder to cancel
+     */
+    fun cancelReminder(context: Context, reminder: Reminder) {
+        (context.getSystemService(ALARM_SERVICE) as? AlarmManager)?.let { alarmManager ->
+            val pendingIntent = pendingIntentForReminder(context, reminder)
+            alarmManager.cancel(pendingIntent)
+            pendingIntent.cancel()
+        } ?: run {
+            logger.warn("Failed to obtain alarm service to cancel all reminders")
+        }
+    }
+
+    /**
+     * Cancels all currently scheduled reminders
+     * This only works if @property allActivePendingIntents returns identical PendingIntents
+     * to the ones used when the notification was originally scheduled
+     * @context can be app or activity
+     */
+    fun cancelAllReminders(context: Context) {
+        enableReceiver(context, false)
+        (context.getSystemService(ALARM_SERVICE) as? AlarmManager)?.let { alarmManager ->
+            allActivePendingIntents(context).forEach {
+                alarmManager.cancel(it)
+                it.cancel()
+            }
+        } ?: run {
+            logger.warn("Failed to obtain alarm service to cancel all reminders")
+        }
+    }
+
+    /**
+     * @context can be app or activity
+     * @param enable if true, the broadcast receiver will be enabled,
+     *               if false, it will be disabled and it won't receive existing pending intents
+     */
+    protected fun enableReceiver(context: Context, enable: Boolean) {
+        val state =
+            if (enable) PackageManager.COMPONENT_ENABLED_STATE_ENABLED
+            else PackageManager.COMPONENT_ENABLED_STATE_DISABLED
+
+        val receiver = ComponentName(context, reminderAlarmReceiver)
+        val pm = context.packageManager
+        pm.setComponentEnabledSetting(receiver, state, PackageManager.DONT_KILL_APP)
+    }
+}
+
+/**
+ * Reminder encapsulates the setup of an alarm notification to be used by the ReminderManager
+ */
+data class Reminder(
+    /**
+     * @property guid uniquely identifying this reminder
+     */
+    val guid: String,
+    /**
+     * @property action notification action
+     */
+    val action: String,
+    /**
+     * @property code notification code
+     */
+    val code: Int,
+    /**
```

Comment:
Thinking about reminders and time zone changes... I think we should schedule/save reminders as a LocalTime (or LocalDateTime) and respond to time zone changes to cancel/reschedule the alarmEpochMillis. What do you think?

```
    <receiver android:name=".TimeZoneChangedReceiver"
        android:enabled="true">
        <intent-filter>
            <action android:name="android.intent.action.TIMEZONE_CHANGED " />
        </intent-filter>
    </receiver>
```

Reply:
This is a good point.  If the timezone changed, the epochMillis would no longer be at the hour/minute that you scheduled it at.  I'll switch over to using LocalDateTime instead, and I will compute the epochMillis before scheduling, based on the user's current timezone.